### ISSUE TITLE
Investigate Digamma op slow-down 

### ISSUE THREAD

**Issue by KalaivaniMCW** [tenstorrent/tt-metal#41936]
Investigate regression in Digamma op perf 

- The digamma LLK was recently merged and improved the performance of the op significantly compared to the composite version 
- But the performance regressed recently and on a quickly analysis points to the SFPI update. 

<img width="1141" height="189" alt="Image" src="https://github.com/user-attachments/assets/3e5ae1f5-4480-4aa1-b3b2-83e690e1518b" />

**Comment by KalaivaniMCW** [comment]
Based on the analysis @Aswinmcw did, 

## Why Only Digamma Regresses with SFPI 7.40.0
**Digamma is the ONLY op using piecewise_rational_eval** — the shared LUT-based evaluator. No other op currently uses this code path.
### What makes digamma unique:
1. **Longest dependency chain of any unary SFPU op:**
   - FP32: degree 10 numerator + degree 9 denominator = **~20 back-to-back SFPMADs** in interleaved Horner
   - BF16: degree 12 + degree 11 = **~24 SFPMADs**
   - Other ops (sigmoid, tanh, gelu, etc.) use shorter hand-tuned polynomial approximations (3-5 FMAs)
2. **Single deferred reciprocal at the very end** (line 285):

   `return numer * sfpu_reciprocal<APPROX_RECIP>(denom);`
   
   The reciprocal uses 2 Newton-Raphson iterations (APPROX_RECIP=false). The compiler must keep both numer and denom live across 20+ FMAs before reaching this point.

3. **2-segment branching** with v_if(x >= lut[SEG]) — conditional writes to numer/denom add register pressure.
### Why SFPI 7.40.0 regresses it:
The s2vFloat16a/b → sFloat16a/b type rename in 7.40.0 changed how the compiler handles scalar float16 types. This likely affected:
- **Register allocation**: The long Horner chain needs many live registers. New type system may cause spills.
- **Instruction scheduling**: SFPI 7.39.0's scheduler may have hidden the reciprocal latency better by overlapping it with earlier FMAs.

**Comment by KalaivaniMCW** [comment]
FYI @nkapreTT

**Comment by KalaivaniMCW** [comment]
cc: @nathan-TT

**Comment by nathan-TT** [comment]
1) I don;t believe the s2vfloat rename changed code generation -- that was certainly the intent.  There were some precursor metal commits to avid behavior that would not be supported.

2) there is no sfp instruction sheduling.

3) SFPU registers cannot be spilled, if you run out, your program does not compile. 

If you'd like further investigation please let me know how to run the case in question from an IRD command line?

**Comment by Aswinmcw** [comment]
Hi @nathan-TT, here are the steps to reproduce:

**1. Test script (`test_digamma_perf.py`):**
```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

input_tensor = ttnn.from_torch(
    torch.rand([1, 1, 32, 32], dtype=torch.bfloat16) + 1.0,
    layout=ttnn.TILE_LAYOUT,
    device=device,
)

output = ttnn.digamma(input_tensor)

ttnn.close_device(device)
```

**2. Run with Tracy profiler to get perf CSV:**
```bash
python3 -m tracy -v -r -p test_digamma_perf.py
```

The CSV with `DEVICE KERNEL DURATION [ns]` will be generated at:
```
$TT_METAL_HOME/generated/profiler/reports/<timestamp>/ops_perf_results_<timestamp>.csv
```

Please test this with both SFPI version 7.39 and 7.40 to see the degradation in perf.

**Comment by nathan-TT** [comment]
is this wormhole or blackhole?

**Comment by Aswinmcw** [comment]
> is this wormhole or blackhole?

Wormhole

**Comment by nathan-TT** [comment]
thanks, I have before and after executables and compiler logs.  It looks like the 7.40 compiler is no longer unrolling (at least) one loop. I had to move some optimizations around for correctness and that's changed the cost calculation at the unroll decision point.  I've not yet determined which loop is affected.

**Comment by nathan-TT** [comment]
It turns out the there's a loop that is just below the unroll limit in the -39 compiler, but in the -40 compiler it is too large (because it has correctly counted instructions).  Looking at adjusting the heuristic

**Comment by nathan-TT** [comment]
Ok, I have a patch:
```
* a27e1054e67 2026-04-14 | Increase unroll insn limit (HEAD -> nsidwell/digamma-41936, origin/nsidwell/digamma-41936) [Nathan Sidwell]
```

This might not be the best way to fix it, but it does seem to produce the same binary with this and the -39 compiler.  and a similar performance (I don;t know what the noise is in a single run).
I'm running CI tests, but perhaps you could test with it as well?

**Comment by KalaivaniMCW** [comment]
Hi @nathan-TT 
 I tested with the change you mentioned  `common_flags += "--param=max-completely-peeled-insns=350 ";`
and the perf is better
<img width="169" height="72" alt="Image" src="https://github.com/user-attachments/assets/8b561170-2c30-4164-a097-f1b3737110b1" />


without the change, the duration is 
<img width="169" height="72" alt="Image" src="https://github.com/user-attachments/assets/f040f57c-1236-4313-8b5d-1f933726ff17" />

**Comment by nathan-TT** [comment]
great, thanks!  I'll raise a PR

**Comment by nathan-TT** [comment]
Beleived fixed by unroll parameter change

**Comment by Aswinmcw** [comment]
Reopening as this fix has been reverted,
cc. @nathan-TT @mateusznowakTT

**Comment by nathan-TT** [comment]
Reverted because:
Well, when you get back up there is also this that happened when you were away:
It looks like your change to adjust the unrolling caused a failure in some tests not fitting anymore. We rolled it back since I didn't think this was something Bob would know how to address "properly" quickly.
https://tenstorrent.slack.com/archives/C05GRJC4J4A/p1776418558957459
GPT-OSS Galaxy demo tests failing on TT_FATAL: Program size (79280) too large for kernel config buffer (70656) on TENSIX https://github.com/tenstorrent/tt-metal/actions/runs/24541428615/job/71748626727[Thread in tt-metal-pipelines](https://tenstorrent.slack.com/archives/C05GRJC4J4A/p1776418558957459) | [Apr 17th](https://tenstorrent.slack.com/archives/C05GRJC4J4A/p1776418558957459) | [View message](https://tenstorrent.slack.com/archives/C05GRJC4J4A/p1776418558957459)

**Comment by nathan-TT** [comment]
Beleived fixed by https://github.com/tenstorrent/tt-metal/pull/43340
